File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/teacher/TeacherServiceImpl.java
package kh.farrukh.sortmapperexamplejava.teacher;

import java.time.LocalDateTime;
import kh.farrukh.sortmapperexamplejava.common.exception.NotFoundException;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherResponseDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherUpdateRequestDTO;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class TeacherServiceImpl implements TeacherService {

  private final TeacherRepo teacherRepo;
  private final TeacherMapper teacherMapper;

  @Override
  public void createTeacher(TeacherCreateRequestDTO requestDTO) {
    var teacher = teacherMapper.toEntity(requestDTO);
    teacherRepo.save(teacher);
  }

  @Override
  public void updateTeacher(Long id, TeacherUpdateRequestDTO requestDTO) {
    var teacher = findTeacher(id);
    teacherMapper.update(teacher, requestDTO);
    teacherRepo.save(teacher);
  }

  @Override
  public void deleteTeacher(Long id) {
    var teacher = findTeacher(id);
    teacher.setDeletedAt(LocalDateTime.now());
    teacherRepo.save(teacher);
  }

  @Override
  public TeacherDetailsResponseDTO getTeacher(Long id) {
    var teacher = findTeacher(id);
    return teacherMapper.toDetailsResponseDTO(teacher);
  }

  @Override
  public Page<TeacherResponseDTO> getTeachers(Pageable pageable) {
    return teacherRepo
        .findAllByDeletedAtIsNull(pageable)
        .map(teacherMapper::toResponseDTO);
  }

  @Override
  public TeacherEntity findTeacher(Long id) {
    return teacherRepo
        .findByIdAndDeletedAtIsNull(id)
        .orElseThrow(() -> new NotFoundException("Teacher not found with id: " + id));
  }

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/teacher/TeacherServiceImpl.kt
package kh.farrukh.sortmapperexample.teacher

import kh.farrukh.sortmapperexample.common.exception.NotFoundException
import kh.farrukh.sortmapperexample.teacher.model.TeacherCreateRequestDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherDetailsResponseDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherResponseDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherUpdateRequestDTO
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import java.time.LocalDateTime

@Service
class TeacherServiceImpl(
    private val teacherRepo: TeacherRepo,
    private val teacherMapper: TeacherMapper
) : TeacherService {

    override fun createTeacher(requestDTO: TeacherCreateRequestDTO) {
        val teacherEntity = teacherMapper.toEntity(requestDTO)
        teacherRepo.save(teacherEntity)
    }

    override fun getTeacherById(id: Long): TeacherDetailsResponseDTO {
        return findTeacher(id)
            .let(teacherMapper::toDetailsResponseDTO)
    }

    override fun getTeachers(pageable: Pageable): Page<TeacherResponseDTO> {
        return teacherRepo
            .findAllByDeletedAtIsNull(pageable)
            .map(teacherMapper::toResponseDTO)
    }

    override fun updateTeacher(id: Long, requestDTO: TeacherUpdateRequestDTO) {
        val teacherEntity = findTeacher(id)
        teacherMapper.update(teacherEntity, requestDTO)
        teacherRepo.save(teacherEntity)
    }

    override fun deleteTeacher(id: Long) {
        val teacherEntity = findTeacher(id)
        teacherEntity.deletedAt = LocalDateTime.now()
        teacherRepo.save(teacherEntity)
    }

    override fun findTeacher(id: Long): TeacherEntity =
        teacherRepo.findByIdAndDeletedAtIsNull(id)
            ?: throw NotFoundException("Teacher not found with id: $id")

}